File: CSE241 - Object Oriented Programming - Homeworks(2022)/hw3(not_graded_but_flawless)/200104004067_AliAsim_Coskun_HW3/driver2.cpp
#include "tetris.h"

using namespace std;
using namespace NS_tetris;

int main()
{
    int random;
    srand(time(NULL));

    int row;
    int column;

    cout << "Please enter the row size of board\n(For example: 7)" << endl;
    cin >> row;
    while(true) //checking validity of row
    {
        if(cin.fail())
        {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "INVALID! Please enter valid value: " << endl;
            cin >> row;
        }
        if(!cin.fail()) break;
    }
    cout << "Please enter the column size of board\n(For example: 10)" << endl;
    cin >> column;
    while(true) //checking validity of column
    {
        if(cin.fail())
        {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "INVALID! Please enter valid value: " << endl;
            cin >> column;
        }
        if(!cin.fail()) break;
    }

    Tetris board(row, column); //creating board with given size
    cout << "Board has been created!" << endl;

    string check_rotation;
    char rotation_direction;
    int rotation_count;

    string check_move;
    char move_direction;
    int move_count;

    string check_type;
    char* type;
    bool over;

    while(true)
    {
        random=9;
        cout << "\n\nPlease select shape of the Tetromino\n(I, J, L, O, S, T, Z)" << endl;
        cout << "If you want to quit, press Q." << endl;
        cin >> check_type;
        type = &check_type[0];

        if((type[0]=='Q' || type[0]=='q') && type[1]=='\0')
        {
            cout << "You want to quit. Goodbye!" << endl;
            break;
        }

        if(type[0]=='R' || type[0]=='r')
        {
            random=rand()%7+1;
        }

        if(((type[0]=='j' || type[0]=='J') && type[1]=='\0') || random==1){
            Tetromino y(Shapes::J);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==0) break;
            board+=y;
            board.animate(y);
        }
        else if(((type[0]=='i' || type[0]=='I') && type[1]=='\0') || random==2){
            Tetromino y(Shapes::I);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==0) break;
            board+=y;
            board.animate(y);
        }
        else if(((type[0]=='o' || type[0]=='O') && type[1]=='\0') || random==3){
            Tetromino y(Shapes::O);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==0) break;
            board+=y;
            board.animate(y);
        }
        else if(((type[0]=='t' || type[0]=='T') && type[1]=='\0') || random==4){
            Tetromino y(Shapes::T);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
                if(temp==0) board+=y;
            }
            if(over==0) break;
            board.animate(y);
        }
        else if(((type[0]=='l' || type[0]=='L') && type[1]=='\0') || random==5){
            Tetromino y(Shapes::L);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
                if(temp==0) board+=y;
            }
            if(over==0) break;
            board.animate(y);
        }
        else if(((type[0]=='s' || type[0]=='S') && type[1]=='\0') || random==6){
            Tetromino y(Shapes::S);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
                if(temp==0) board+=y;
            }
            if(over==0) break;
            board.animate(y);
        }
        else if(((type[0]=='z' || type[0]=='Z') && type[1]=='\0') || random==7){
            Tetromino y(Shapes::Z);
            over=(board+=y);
            if(over==0) break;

            cout << "rotation direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_rotation;
            while(!((check_rotation[0]=='R' && check_rotation[1]=='\0') || (check_rotation[0]=='L' && check_rotation[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_rotation;
            }
            rotation_direction=check_rotation[0];
            cout << "rotation count?\n(Enter a number): ";
            cin >> rotation_count;
            while(true) //checking validity of rotation count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> rotation_count;
                }
                if(!cin.fail()) break;
            }
            cout << "move_direction?\n(Enter 'L' for Left, 'R' for Right): ";
            cin >> check_move;
            while(!((check_move[0]=='R' && check_move[1]=='\0') || (check_move[0]=='L' && check_move[1]=='\0')))
            {
                cout << "INVALID! Please try again: " << endl;
                cin >> check_move;
            }
            move_direction=check_move[0];
            
            cout << "move count?\n(Enter a number): ";
            cin >> move_count;
            while(true) //checking validity of move count
            {
                if(cin.fail())
                {
                    cin.clear();
                    cin.ignore(numeric_limits<streamsize>::max(), '\n');
                    cout << "INVALID! Please enter valid value: " << endl;
                    cin >> move_count;
                }
                if(!cin.fail()) break;
            }
            board.setMove(move_direction, move_count);
            int temp=rotation_count;

            while(temp>0)
            {
                over=(board+=y);
                if(over==0) break;
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
                if(temp==0) board+=y;
            }
            if(over==0) break;
            board.animate(y);
        }
        else{
            cout << "Invalid Input!" << endl;
        }
    }
    if(over==0)
    {
        cout << "GAME OVER!" << endl;
    }

    return 0;
}

File: CSE241 - Object Oriented Programming - Homeworks(2022)/hw7(85)/200104004067_AliAsim_Coskun_HW7/driver2.java
import java.util.Scanner;
import java.util.Random;
import java.util.TimeZone;

/**
 * This class is the driver class for the Tetris game, which allows the user to play the game and test the functionality of the Tetris and Tetromino classes.
 * The driver class uses a Scanner object to take input from the user and a Random object to generate random values for the game.
 * It creates a Tetris board and Tetromino objects, and uses the various methods from the Tetris and Tetromino classes to play the game.
 * The user can select a shape of Tetromino, rotate and move it on the board, and see the animation of the Tetromino falling and filling the board.
 * The user can also quit the game by pressing the 'Q' key.
 * The class also includes 5 different test runs to test the functionality of the game with different board sizes and Tetromino shapes.
*/
public class driver2 {
    /**
     * The main method of the driver2 class. It runs 5 different test cases,
     * each with different board sizes and Tetromino shapes.
     *
     * @param args command line arguments (not used)
     */
    public static void main(String[] args) {

    System.out.println("DRIVER2\n\n\n\n");

    int random;
    Random rand = new Random();
    rand.setSeed(System.currentTimeMillis());

    int row;
    int column;

    Scanner sc = new Scanner(System.in);

    System.out.println("Please enter the row size of board\n(For example: 7)");
    row = sc.nextInt();

    System.out.println("Please enter the column size of board\n(For example: 10)");
    column = sc.nextInt();

    Tetris board = new Tetris(row, column);
    System.out.println("Board has been created!");

    char rotation_direction;
    int rotation_count;

    char move_direction;
    int move_count;

    boolean over=true;
    char check;

    Random rand2 = new Random();
    rand2.setSeed(System.currentTimeMillis());
    int random_for_rotate;

    Random rand3 = new Random();
    rand3.setSeed(System.currentTimeMillis());
    int random_for_move;
    
    Random rand4 = new Random();
    rand4.setSeed(System.currentTimeMillis());
    int random_for_rotation_direction;
    
    Random rand5 = new Random();
    rand5.setSeed(System.currentTimeMillis());
    int random_for_move_direction;

    if(sc.hasNextLine()){
        sc.nextLine();
    }
    while(true) {
        random=9;
        System.out.println("\n\nPlease select shape of the Tetromino\n(I, J, L, O, S, T, Z)\nor 'R' for Random.");
        System.out.println("If you want to quit, press Q.");
        check = sc.nextLine().charAt(0);

        random_for_rotate = rand.nextInt(4);
        random_for_move = rand.nextInt(column/2);
        random_for_rotation_direction = rand.nextInt(2);
        random_for_move_direction = rand.nextInt(2);

        // System.out.println(check);
        // check='J';

        if(check=='Q') {
            System.out.println("You want to quit. Goodbye!");
            break;
        }

        if(check=='R') {
            random = rand.nextInt(7) + 1;
        }

        if(check=='J' || random==1) {
            Tetromino y = new Tetromino(Shapes.J);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='L') || random==2) {
            Tetromino y = new Tetromino(Shapes.L);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='O') || random==3) {
            Tetromino y = new Tetromino(Shapes.O);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='S') || random==4) {
            Tetromino y = new Tetromino(Shapes.S);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='T') || random==5) {
            Tetromino y = new Tetromino(Shapes.T);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='Z') || random==6) {
            Tetromino y = new Tetromino(Shapes.Z);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else if((check=='I') || random==7) {
            Tetromino y = new Tetromino(Shapes.I);
            over = (board.addTetromino(y));
            if(over == false) break;

            if(random_for_rotation_direction==0) rotation_direction ='L';
            else rotation_direction ='R';
            rotation_count=random_for_rotate;

            if(random_for_move_direction==0) move_direction='L';
            else move_direction='R';
            move_count=random_for_move;
            
            board.setMove(move_direction, move_count);
            int temp=rotation_count;
            while(temp>0) {
                over = (board.addTetromino(y));
                if(over==false) {
                    break;
                }
                y.rotate(rotation_direction, y.getShapeRow(), y.getShapeColumn());
                temp--;
            }
            if(over==false) {
                break;
            }
            board.addTetromino(y);
            board.animate(y);
        }
        else{
            System.out.println("Invalid Input! \n");
        }
    }
    if(over==false)
    {
        System.out.println("GAME OVER! \n");
    }

    sc.close();
    }
}